cmake/add_enclave.cmake

```diff
@@ -0,0 +1,62 @@
+# Copyright (c) Microsoft Corporation. All rights reserved.
+# Licensed under the MIT License.
+#
+# Helper function to sign an enclave binary.
+#
+# Usage:
+#
+#	add_enclave(<TARGET target> <SOURCES sources> <CONFIG config>)
+#
+# The target must already exist and should be a shared object (an
+# enclave).
+#
+# Given <target> and <signconffile>, this function adds custom
+# commands to generate a signing key and call `oesign` to sign the
+# target, resulting in `<target>.signed.so`. It also adds
+# `<target>_signed` as an imported target so that it can be referenced
+# later in the CMake graph.
+# TODO: (1) Replace the name guessing logic.
+# TODO: (2) Setup the dependency using `${BIN}_signed` instead of the
+# default custom target.
+function(add_enclave)
+
+   set(options)
+   set(oneValueArgs TARGET CONFIG KEY)
+   set(multiValueArgs SOURCES)
+   cmake_parse_arguments(ENCLAVE "${options}" "${oneValueArgs}"
+                          "${multiValueArgs}" ${ARGN})
+   add_executable(${ENCLAVE_TARGET} ${ENCLAVE_SOURCES})
+   
+   if(NOT ENCLAVE_CONFIG)
+      return()
+   endif()
+
+  # Generate the signing key.
+  if(NOT ENCLAVE_KEY)
+     add_custom_command(OUTPUT ${ENCLAVE_TARGET}-private.pem
+       COMMAND openssl genrsa -out ${ENCLAVE_TARGET}-private.pem -3 3072)
+     set(ENCLAVE_KEY  ${CMAKE_CURRENT_BINARY_DIR}/${ENCLAVE_TARGET}-private.pem)
+  endif()
+
+  # TODO: Get this name intelligently (somehow use $<TARGET_FILE> with
+  # `.signed` injected).
+  set(SIGNED_LOCATION ${CMAKE_CURRENT_BINARY_DIR}/${ENCLAVE_TARGET}.signed.so)
+
+  # Sign the enclave using `oesign`.
+  if(ENCLAVE_CONFIG)
+    add_custom_command(OUTPUT ${SIGNED_LOCATION}
+      COMMAND oesign $<TARGET_FILE:${ENCLAVE_TARGET}> ${ENCLAVE_CONFIG} ${ENCLAVE_KEY} 
+      DEPENDS oesign ${ENCLAVE_TARGET} ${ENCLAVE_CONFIG} ${ENCLAVE_KEY}
+      WORKING_DIRECTORY ${CMAKE_CURRENT_SOURCE_DIR})
+  endif()
+
+  # Import the generated signed enclave so we can reference it with
+  # `$<TARGET_FILE>` later.
+  add_library(${ENCLAVE_TARGET}_signed SHARED IMPORTED GLOBAL)
+  set_target_properties(${ENCLAVE_TARGET}_signed PROPERTIES
+    IMPORTED_LOCATION ${SIGNED_LOCATION})
+
+  # Add a custom target with `ALL` semantics so these targets are always built.
```

Comment:
Probably add a note that this is kind of like a "fake" Makefile target. It exists in the graph to trigger a build, but is not otherwise used.

Whereas the imported target is also fake, it exists in the graph _so that it can be used elsewhere_.

Reply:
And specifically that the custom target is required in order to set a file-level dependency on `${SIGNED_LOCATION}`, which you cannot (sadly) otherwise do with the imported executable. I think I left some notes on this in my commits, but they got disappeared. If you want, I can track them down and paste the notes here. I think they're worth keeping.

Reply:
Cause this logic is weird 🤣